### PR TITLE
GOINGTOSAI: SWI-4527 - Prepare saibuffer.py file for upstreaming

### PR DESCRIPTION
**Pull request by mmysle**

### CODE REVIEW

**Comment by mmysle** [comment]
The IngressPriorityGroupScale test was removed due to Tofino-specific reason. The test case was marked as internal.

**Comment by ravi861** [comment]
@mickeyspiegel @mkbalani Can one of you please review this?

File: ptf/saibuffer.py [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+        self.assertEqual(status, SAI_STATUS_SUCCESS)
+
+        print("Buffer pool size:", self.buf_size)
+        print("Buffer profile reserved_buffer_size:", self.reserved_buf_size)
+        print("Buffer profile shared_static_th:", self.reserved_buf_size)
+
+        self.sendVerify(expected_drops=0, verify_reserved_buffer_size=False)
+        self.clearVerify()
+
+        self.pkt_len = 1500
```
Comment (ustachowiak) [inline ptf/saibuffer.py:235]: Maybe do it without subtracting?

File: ptf/saibuffer.py [hunk 2]
```diff
@@ -0,0 +1,671 @@
...
+            stats["SAI_INGRESS_PRIORITY_GROUP_STAT_PACKETS"], 0)
+        self.assertEqual(
+            stats["SAI_INGRESS_PRIORITY_GROUP_STAT_BYTES"], 0)
+        self.assertEqual(
+            stats["SAI_INGRESS_PRIORITY_GROUP_STAT_WATERMARK_BYTES"], 0)
+        self.assertEqual(
+            stats["SAI_INGRESS_PRIORITY_GROUP_STAT_DROPPED_PACKETS"], 0)
+
+    def runTest(self):
+        print()
```
Comment (ustachowiak) [inline ptf/saibuffer.py:217]: Are these empty prints at the beginning of functions needed? Same for the other functions.

File: ptf/saibuffer.py [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+        print("Create {} ingress buffer pools".format(ingr_pool_num))
+        for i in range(ingr_pool_num):
+            print("Create ingr_pool", i)
+
+            ingr_pool = sai_thrift_create_buffer_pool(
+                self.client, type=SAI_BUFFER_POOL_TYPE_INGRESS,
+                size=self.buf_size + i)
+            self.assertGreater(ingr_pool, 0)
+            self.ingr_pools.append(ingr_pool)
+        print("OK")
```
Comment (ustachowiak) [inline ptf/saibuffer.py:460]: "OK" print doesn't sound good

File: ptf/saibuffer.py [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+        print("Create {} egress buffer pools".format(egr_pool_num))
+        for i in range(egr_pool_num):
+            print("Create egr_pool", i)
+
+            egr_pool = sai_thrift_create_buffer_pool(
+                self.client, type=SAI_BUFFER_POOL_TYPE_EGRESS,
+                size=self.buf_size + i)
+            self.assertGreater(egr_pool, 0)
+            self.egr_pools.append(egr_pool)
+        print("OK")
```
Comment (ustachowiak) [inline ptf/saibuffer.py:471]: Same here

File: ptf/saibuffer.py [hunk 5]
```diff
@@ -0,0 +1,671 @@
...
+                shared_static_th=self.profile_buf_size,
+                threshold_mode=SAI_BUFFER_PROFILE_THRESHOLD_MODE_STATIC)
+            self.assertGreater(profile, 0)
+            self.ingr_profiles.append(profile)
+
+            ipg = sai_thrift_create_ingress_priority_group(
+                self.client, port=self.port0, index=i, buffer_profile=profile)
+            self.assertGreater(ipg, 0)
+            self.ipgs.append(ipg)
+        print("OK")
```
Comment (ustachowiak) [inline ptf/saibuffer.py:489]: And here

File: ptf/saibuffer.py [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+                reserved_buffer_size=self.profile_buf_size,
+                shared_static_th=self.profile_buf_size,
+                threshold_mode=SAI_BUFFER_PROFILE_THRESHOLD_MODE_STATIC)
+            self.assertGreater(profile, 0)
+            self.egr_profiles.append(profile)
+
+            status = sai_thrift_set_queue_attribute(
+                self.client, self.queues[i], buffer_profile_id=profile)
+            self.assertEqual(status, SAI_STATUS_SUCCESS)
+        print("OK")
```
Comment (ustachowiak) [inline ptf/saibuffer.py:514]: And here

File: ptf/saibuffer.py [hunk 7]
```diff
@@ -0,0 +1,671 @@
...
+        self.assertGreater(self.dscp_to_tc_map, 0)
+
+        status = sai_thrift_set_port_attribute(
+            self.client, self.port0, qos_dscp_to_tc_map=self.dscp_to_tc_map)
+        self.assertEqual(status, SAI_STATUS_SUCCESS)
+
+        status = sai_thrift_set_port_attribute(
+            self.client, self.port1, qos_dscp_to_tc_map=self.dscp_to_tc_map)
+        self.assertEqual(status, SAI_STATUS_SUCCESS)
+        print("OK")
```
Comment (ustachowiak) [inline ptf/saibuffer.py:561]: And here

File: ptf/saibuffer.py [hunk 8]
```diff
@@ -0,0 +1,671 @@
...
+        self.prio_to_ipg_map = sai_thrift_create_qos_map(
+            self.client, type=SAI_QOS_MAP_TYPE_PFC_PRIORITY_TO_PRIORITY_GROUP,
+            map_to_value_list=qos_map_list)
+        self.assertGreater(self.prio_to_ipg_map, 0)
+
+        status = sai_thrift_set_port_attribute(
+            self.client, self.port0,
+            qos_pfc_priority_to_priority_group_map=self.prio_to_ipg_map)
+        self.assertEqual(status, SAI_STATUS_SUCCESS)
+        print("OK")
```
Comment (ustachowiak) [inline ptf/saibuffer.py:591]: And here

File: ptf/saibuffer.py [hunk 9]
```diff
@@ -0,0 +1,671 @@
...
+
+        self.tc_to_q_map = sai_thrift_create_qos_map(
+            self.client, type=SAI_QOS_MAP_TYPE_TC_TO_QUEUE,
+            map_to_value_list=qos_map_list)
+        self.assertGreater(self.tc_to_q_map, 0)
+
+        status = sai_thrift_set_port_attribute(
+            self.client, self.port1, qos_tc_to_queue_map=self.tc_to_q_map)
+        self.assertEqual(status, SAI_STATUS_SUCCESS)
+        print("OK")
```
Comment (ustachowiak) [inline ptf/saibuffer.py:605]: And here

File: ptf/saibuffer.py [hunk 10]
```diff
@@ -0,0 +1,671 @@
...
+
+    def setUp(self):
+        super(BufferStatistics, self).setUp()
+
+        self.tx_cnt = 10000
+        self.pkt_len = 700
+        self.reserved_buf_size = 1400
+        self.buf_size = self.reserved_buf_size * 1000
+        self.sleep_time = 2
+        self.pkt = simple_udp_packet(
```
Comment (ustachowiak) [inline ptf/saibuffer.py:42]: Maybe do it without subtracting?

File: ptf/saibuffer.py [hunk 11]
```diff
@@ -0,0 +1,671 @@
...
+        super(Forwarding, self).__init__(port_num=3)
+
+    def setUp(self):
+        super(Forwarding, self).setUp()
+
+        self.tx_cnt = 100
+        self.pkt_len = 1500
+        self.reserved_buf_size = 3200
+        self.buf_size = self.reserved_buf_size * 1000
+        self.pkt = simple_udp_packet(
```
Comment (ustachowiak) [inline ptf/saibuffer.py:276]: Maybe do it without subtracting?

**Review by ustachowiak — APPROVED** [review]

File: ptf/saibuffer.py (hunk 2 above)
Comment (mmysle) [reply ptf/saibuffer.py:217]: For readability reason during test execution - maybe more Pythonic than "\n" - a matter of taste.

File: ptf/saibuffer.py (hunk 1 above)
Comment (mmysle) [reply ptf/saibuffer.py:235]: Documented why.

**Comment by mmysle** [comment]
SAI PTF buffer module was uploaded for external review: https://github.com/opencomputeproject/SAI/pull/1367

**Pull request closed without merging**